### DISCUSSION TITLE
more rules

### DISCUSSION THREAD

**Discussion by zmechu** [arendst/Tasmota#15832] [Ideas]
Hello!

Tasmota has 3 rules available. Of course rules can be multi-line and it can be put more commands into one rule, but is there a way to change number of rules? For example 10, or add sth like #MORE_RULES_COMPILE_OPTION 10?

**Comment by sfromis** [comment]
Actually not only 3 rules. Tasmota has 3 rule **sets** available, but each of those can contain many rules (`on...endon`). With 3 buffers of ½ KB available, and compression kicking in, you can generally have over 3 KB of rule text.

**Reply by zmechu** [reply]
Yeah, exactly, but in my opinion it could be more convenient to have more short rules than only few long. For example: one rule for buttons, one rule for LCD, one for timers, one for sensors, one for system state and so on.

It is hard to make changes/display/show 3kB rule and its increase error probability (logic, syntax and so on), with harder debugging...

**Reply by sfromis** [reply]
I can much recommend preparing/saving your rule sets in an editor, and only update (via MQTT, copy/paste or whatever) as quick one-step operations. This way, you can still have one rule `on...endon` per line.

Due to the architecture of Tasmota, it is rather unlikely that the number of rule sets will be changed.

**Reply by zmechu** [reply]
I do such, using Notepad and copy+paste. But it is still really hard to cover many `on..endon`  code lines.

I hope there is way to increase number of rules. Tasmota is so powerful, that adding 0,5kB of code will be piece a cake :)

**Reply by sfromis** [reply]
Is it hard to cover many lines in notepad, and do **one single** copy/paste? Not one copy/paste per rule in the set, but one multiline copy/paste for the whole set?

**Reply by sfromis** [reply]
Useless example, having 3 lines 
```
rule1
on Event#a do var1 a endon
on Event#b do var1 b endon
```
becomes one line when pasted into the console command field:
`rule1 on Event#a do var1 a endon on Event#b do var1 b endon`

**Reply by zmechu** [reply]
My idea is such: 3 rules are fine, but more rules will be more convenient. Of course I can put and combine many of sets into one rule - using notepad, using copy-paste, managing many sets into one etc, etc, etc. 

But in my opinion it will be nice option to have possibility to use many small rules in a place of three huge. I have three sensors - it will be more easy to use one rule per sensor than put one set into 0.5kB command.

For example, now I have:
```
Rule1 ON BH1750#Illuminance>%var1% DO Backlog...
Rule1 + ON BH1750#Illuminance<%var2% DO Backlog...
rule1 + ON tele-SHT3X#Temperature DO DisplayText...
rule1 + ON tele-SHT3X#DewPoint DO DisplayText
rule1 + ON Time DO...
rule1 + ON System#Boot DO...
rule1 + ON Switch2#State DO...
```
and so on. It works, but there is mixed sets of many sensors, timer, system and it is only first rule :-)
It could be more easy to manage rules using:
```
Rule1 ON BH1750#Illuminance>%var1% DO Backlog...
Rule1 + ON BH1750#Illuminance<%var2% DO Backlog...

rule2 ON tele-SHT3X#Temperature DO DisplayText...
rule2 + ON tele-SHT3X#DewPoint DO DisplayText

rule3 ON Time DO...

rule4 ON System#Boot DO...

rule5 ON Switch2#State DO...
```
so, one rule for one sensor, second rule for second sensor, third rule fir timer, fourth for system, fifth for switches, sixth rule to to rule them all ;-)

My idea is to add functionality to manage/use/view rules more easy.

**Reply by sfromis** [reply]
I do not see much management difficulty... You could just order the rules the way you like, including blank lines. And that `Rule1 +` is a laborious way - it is much simpler without all those +, like I posted: **One**, just one copy/paste taking multiple lines. You are making it much harder than what I posted, needing multiple copy/paste.

Anyway, expansion not going to happen. For more involved needs, ESP32 and Berry much recommended over rules.

**Comment by Jason2866** [comment]
Thank you for sharing your idea. It will not implemented since it is not a small code change and it will result in a bigger firmware size. There are good alternatives to solve how to structure rules.

**Reply by zmechu** [reply]
I mean - 3 rules "built-in" are fine. Mabybe #MORE_RULES not included in precompiled binaries, but possibility to define more with "every rule add 50kB in firmware". If firmware filesize and number of rules are fine that great, if want more be warned :-)

As alternatives - do you mean Scripting (https://tasmota.github.io/docs/Scripting-Language/)?

**Reply by sfromis** [reply]
Scripting can be used instead of rules, but that's not about "how to structure rules".

I already explained how easy that is, instead of your very laborious way with many commands and using + to concatenate line by line. Did you understand the point of what I explained? You just seemed to ignore it?

And barbudor also explained how TDM makes editing easier than the Tasmota console.

**Reply by zmechu** [reply]
I got it, im using "+" just for manage rules in logic way.

TDM means Tasmota Device Manager?

**Reply by sfromis** [reply]
Well, that + makes it much more laborious to manage, and you asked for making it easier to manage. One single copy/paste can do it for a whole rule set, nicely formatted on multiple lines in the editor.

Yes, TDM is [Tasmota Device Manager](https://tasmota.github.io/docs/Tasmota-Device-Manager/).

**Comment by barbudor** [comment]
TDM allows to edit multi-line rules easily

**Comment by AleXSR700** [comment]
I saw that Jason does not want to increase the number of rules or add a compile option, but I would like to make another case for more rule sets.
Yes, it is possible to have multiple rules in one rule set and this allows for a lot of different scenarios.

But, there are also a lot of cases where you need one rule set per option.
Mainly when you want to work with thresholds and want to prevent the rule from triggering over and over.
Since there is no AND / OR condition available by default, additional rule sets would use up less firmware space and it would be possible to limit the length of the rules based on the number of rules.

So if, e.g. the sum of characters for 3 rules was now limited to e.g. 500 due to storage space, then when you use 5 rule sets each rule set could have 100 characters. If you use 10 rule sets, each would only have 50 characters available.

So it would be pretty cool to be able to use more rule sets with smaller rules.

Just my 2 cents in case anybody wants to implement something like this :)

P.S.: the space needed to make simple tasks happen within on rule set by using variables will be larger than the space needed using multiple rule sets. Because sometimes you have to get really creative to get around the rule set limit and that takes a lot of extra chars.

**Reply by barbudor** [reply]
Hi @AleXSR700
I'm sorry I don't get your point about having more shorter rules-set than the existing larges rules-set
I see only 1 technical reason to split rules to a different rules-set, it is when I want to enable "once". In that case I have to dedicate 1 rules-set to hold all triggers that need the once mode.
Beside, I feel that smaller rules-set will reduce compression efficiency.

Can you please elaborate your position?
Have you ever reached the max capacity of all 3 rules-set despite compression?

**Reply by AleXSR700** [reply]
Well, I see thresholds as a reason.
If you try to launch different actions with different thresholds, you need to use `ON BREAK`.
This limits the amount of rules you can combine.

Just as an example:
If you want different powerdelta or teleperiods for different absolute energy#power, you do something like
```
On energy#power[1]>1000 Do Backlog teleperiod 600, powerdelta 120 Break
On energy#power[1]>100 Do Backlog teleperiod 600, powerdelta 110 Break
On energy#power[1]<10 Do Backlog teleperiod 3600, powerdelta 102 Endon
```
You will then need a second rule set for the second power output (e.g. on Shelly 2.5).
```
On energy#power[2]>1000 Do Backlog teleperiod 600, powerdelta 120 Break
On energy#power[2]>100 Do Backlog teleperiod 600, powerdelta 110 Break
On energy#power[2]<10 Do Backlog teleperiod 3600, powerdelta 102 Endon
```
They cannot be combined due to the need for `break`.
If you now want to do anything else related to output 1 & 2, you will need 2 more rule sets and not just 1 more.

The `break` is a cool way of simulating if else but it does not give the same freedom. So you may quickly run into the limitations of "only" 3 rule sets on device with e.g. two outputs.

**Reply by sfromis** [reply]
If what you want is not rule sets but IF/ELSE, do use a build including support for this.

**Reply by AleXSR700** [reply]
Well, there is a reason why IF ELSE is not default and rule sets are, despite if else being much more practical.

So I would guess that additional rule sets might also need less space than if else?

**Reply by barbudor** [reply]
@AleXSR700 are you sure that you cannot combine them?
Normally BREAK applies only to a given trigger so you can combine multiple BREAK/ENDON sequences on the same rule-set
But I must agree that your case is a bit special due to the array. I'm not sure what Tasmota considers as the trigger and as the condition. Is the array index part of the trigger or the condition?

Regarding flash size, more rules set would use more 'data' bytes while 'if/else' requires more 'code' bytes. But at the end it's the same flash.
If/else requires 4Kb of code
How much bytes would you want for your additional rules sets?

**Reply by sfromis** [reply]
> so you can combine multiple BREAK/ENDON sequences on the same rule-set

Yes, but... Only if those are not from the same source. When processing fields coming from the same JSON payload, getting to a `break` will stop evaluation of the whole rule set.

Normally, it is not too much of a problem with "moderate" increases in flash usage for code, hence my pointer about using IF/ELSE when this is the real use case, but with 4KB settings structure being a hard limit with the Tasmota architecture, no expansion here.

However, it would not need an expansion to, say, have 6 rule sets of each only 256 bytes, instead of 3 x 512. While the numbers do have constants, there are several places in the code depending on it being 3 rule sets, making it much more than a parameter change to redistribute the 1.5 KB to smaller rule sets. And smaller rule sets would indeed make the compression less effective.

**Reply by AleXSR700** [reply]
Yes, the issue is that they sometimes rely on the same trigger.

As I said, it is rare but often you have to bend over backwards to make something work in one rule set because you would otherwise run out of them.

IF at any point someone were to increase the number of rule sets, then maybe the code would not have static rule set sizes but instead use a sum. So if you use x chars in rule set 1, you will have total-x available for the rest.
And essentially you could have e.g. 10 rule sets but they would then be very limited in size (with size not being a constant but depending on "consumed" chars. Taking into account also compression.
That would ensure backward compatibility (in case someone uses huge rule sets right now he will simple be limited to less rule set because he/she already consumed their quota).

**Reply by sfromis** [reply]
Well, do start using the realistic solution of using IF/ELSE/ENDIF. Mostly it is no problem adding this feature, obviously depending on what else you add of features.

**Reply by AleXSR700** [reply]
> Well, do start using the realistic solution of using IF/ELSE/ENDIF. Mostly it is no problem adding this feature, obviously depending on what else you add of features. 

Well, the primary disadvantage is the need to compile every dev/master version yourself rather than batch updating all 20 Shellies OTA in one go ;)
If possible, stock is always the best option for maintainability

**Reply by barbudor** [reply]
Is your need for more rules limited to devices with 4MB so filesystem could be used to store those additional rules ?
- either with a dedicated file (rules.dat)
- or using autoexec to load RAM-only Rules with index >3

I have not looked at the code in details for now, so I have no idea if a direction is better/easier than another


> stock is always the best option for maintainability

I prefer my own binaries as I have specific options burned-in: Wifi, MQTT, passwords, auto-config rules... things that insure that even with a "hard-reset" of settings, a device with always get back in the network in the expected state without user interaction

**Reply by AleXSR700** [reply]
This applies to any device.
Primarily ESP8266, which, in the case of Shelly, have 2 MB flash, if I am not mistaken.

Out of curiosity: would it not make more sense to remove rules (save 1.5 kB) by default and add if/else (only 2.5 kB more) as default?
If someone really needs those 2.5 kB, he/she could then deactivate during compiling (as he/she is compiling anyway).

I am just asking because I think if/else is far more common and far easier to use with far less code.
Instead of a 2 liner for a range (one line for > and an identical one for <) you would have one line with two conditions.
And you would not run into the problems with using Break etc. and having to split over multiple rule sets.

I understand the need for Berry, but I am not sure if not maybe if/else would be the future and rules the past.

**Reply by sfromis** [reply]
There exists an "unofficial" build for ESP8266 including "full rules", but as an "experimental" build, content might change without notice. Still, usually works, depending on other needs.
https://github.com/tasmota/install/blob/firmware/firmware/unofficial/tasmota-fullrules.bin.gz
(Note that OTA for ESP8266 cannot use the URL directly, only after downloading)

**Reply by sfromis** [reply]
> Out of curiosity: would it not make more sense to remove rules (save 1.5 kB) by default and add if/else (only 2.5 kB more) as default?

This does not make sense, as IF/ELSE/ENDIF is an extension to rules. And the 1.5KB is settings space for storing rules, while the 4KB for IF/ELSE/ENDIF is code space outside of settings.

**Reply by AleXSR700** [reply]
> This does not make sense, as IF/ELSE/ENDIF is an extension to rules.

So if/elseif uses parts of the code of rules... hmmm...

OTA is kind of the point though. Compiling is not really the issue :)

I will check out the setup of a personal OTA server on a RPI and see if I can maybe switch to if/else without loosing the ability for OTA updates.

**Reply by sfromis** [reply]
No, IF/ELSE/ENDIF does not "use parts of the code of rules", it is the other way around, it is an extension of the capabilities of rules.

It is simple to do URL-based OTA with your own build, just set up a small http server while you need to do upgrades.

**Reply by AleXSR700** [reply]
If if/elseif does not use the code of rules, then what do rules add that if/elseif cannot do?

**Reply by sfromis** [reply]
It does not make sense to talk about IF/ELSE/ENDIF separately, it cannot even exist outside of already having rules included. What I said was that it does not use **parts** of the code of rules. You appear to think of IF/ELSE/ENDIF as some sort of thing "instead of rules". It is not, but additional features for rules.

You could argue that it ises **all** of rules, but still not logical to talk about IF/ELSE/ENDIF "using" rules code at all, when it is instead rules using the IF/ELSE/ENDIF feature.

**Reply by AleXSR700** [reply]
Okay, what I meant with using parts of the code is that it relies on the code or rules to exist, i.e. it is an extension of the code rather than a standalone feature.

I have run into character limits for my Shelly 3EM rules after switching it from if/elseif. So if it is possible to adjust the limitations of rules, it would really be nice to have a total size rather than individual allocations.
That way you could have rule sets that belong together but are large and rule sets that cannot be combined but are small.

**Reply by AleXSR700** [reply]
So, I have been compiling my firmwares now and can OTA update them locally.

But I am still running into the issue of 511 chars per rule set and hence a total of 1533 chars for all 3 rule sets.

Is there an alternative?
I see that the UFS that would allow me to use autoexec.bat needs at least 2 MB. Although I am not sure if it really needs 2 or if the minimum setting for the UFS is just 1 MB (so if I could somehow build an image with less than 1 MB UFS.
Something like `board = esp8266_1M0.1M ` aka 1 MB flash with 0.1 MB UFS.

@barbudor , you mentioned a rules.dat. I would not mind having all rules in one big file if that would be a workaround.
Right now my firmware 643 KB with 360 KB free. So that should easily be enough for a rules.dat if the firmware can somehow support reading such a file.

**Reply by sfromis** [reply]
With compression, the 511 chars per rule can hold way more, typically well above 1 KB.

The alternative is to use ESP32 and Berry. Or cut down on your micro-management tendencies.

**Reply by AleXSR700** [reply]
Cannot use Berry on all my esp8266 devices.
And I don't consider it micro management because it saves me from my database exploding.

I will see if compression solves it. Makes it a bit more tricky because for hat to work you need to create the rule in steps, i.e.
```
rule1 blah blah
rule1 +more blah blah
```
because you cannot just add e.g. 1500 chars in one go.

So setting the rules via mqtt needs to be adjusted into multi liners.

**Comment by barbudor** [comment]
@AleXSR700 You may want to try that [PR19348](https://github.com/arendst/Tasmota/pull/19438) by pulling my branch https://github.com/barbudor/Tasmota/tree/extended_rules

It requires filesystem to save the additional rules so it is not officially supported on 1MB devices.

However I have test version that you can try at your own risks:
Build with the following environment by adding the 2 attached files in `Tasmota/boards/`
```
[env:tasmota-rulesext1m]
build_flags             = ${env.build_flags} -DUSE_RULES_EXTENDED
board                   = esp8266_1M32
```
I'm not sure how safe it is. The filesystem may get corrupted during OTA with a large file

Don't forget that to move from a build without filesystem to a build with filesystem, you need to fully erase and serial-reflash the device (don't forget to save the configuration before)

[esp8266_1M32.zip](https://github.com/arendst/Tasmota/files/12504561/esp8266_1M32.zip)

Note that there is no guarantee that Theo will accept that PR

**Reply by AleXSR700** [reply]
Thank you @barbudor , that looks really interesting 👏

Why is filesystem needed? It reads as if only a few more bytes are needed. Why can't the rules be saved in the free 360 kB of a 1 MB system?

**Reply by barbudor** [reply]
5 additional rules of 512 bytes each
Do the math, that's 2560 bytes
There's no way to store that anywhere else than in file system as we won't reorganize the map of Tasmota

**Reply by AleXSR700** [reply]
I did, but am I then misreading the information?

```
Flash Size 1024 KB
Program Flash Size 1024 KB
Program Size 643 KB
Free Program Space 360 KB
Free Memory 24.6 KB
```
And that info is from a build with If/Else enabled. So additional ~3 kB should fit, should it not?

**Reply by sfromis** [reply]
No. The settings data structure is 4KB fixed, with many different things, not just rule sets. Expanded number of rule sets needs to be stored somewhere, not just assuming that "free space" is available that way, as mechanisms need to exist for reserving space. Using a file system is a good way of having persistent space for such.

**Comment by jgoldwein** [comment]
this is how rule1 appears on the tasmota console.  How do I get rid of all of the extraneous white space?


15:20:53.532 CMD: rule1
15:20:53.541 MQT: merion/kitchen/coffeemaker/stat/tasmota_9E2681/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":322,"Free":189,"Rules":"on System#boot do backlog power1 0;power2 1;power3 1;var4 full;buzzer 1,1,1;rule2 1;rule3 1;delay 50 endon                     on wifi#connected do backlog buzzer 2,1,1;power2 0;power3 0 endon \t\t    on button1#State do power1 2 endon \t\t    on power1#state=1 do buzzer -1,3,7 endon \t\t    on power1#state=0 do buzzer 0 endon"}}

**Reply by barbudor** [reply]
Retype it

**Reply by sfromis** [reply]
You using tabs may look like a lot of white space, but that's likely just how it is presented. I'd still prefer to avoid tabs, as they are not relevant to Tasmota.
